src/ripple/protocol/impl/STObject.cpp

```diff
@@ -694,67 +694,61 @@ bool STObject::operator== (const STObject& obj) const
     return true;
 }
 
-void STObject::add (Serializer& s, bool withSigningFields) const
+void STObject::add (Serializer& s, WhichFields whichFields) const
 {
-    std::map<int, STBase const*> fields;
-    for (auto const& e : v_)
-    {
-        // pick out the fields and sort them
-        if ((e->getSType() != STI_NOTPRESENT) &&
-            e->getFName().shouldInclude (withSigningFields))
-        {
-            fields.insert (std::make_pair (
-                e->getFName().fieldCode, &e.get()));
-        }
-    }
+    // Depending on whichFields, signing fields are either serialized or
+    // not.  Then fields are added to the Serializer sorted by fieldCode.
+    std::vector<STBase const*> const
+        fields {getSortedFields (*this, whichFields)};
 
     // insert sorted
-    for (auto const& e : fields)
+    for (STBase const* const field : fields)
     {
-        auto const field = e.second;
-
         // When we serialize an object inside another object,
         // the type associated by rule with this field name
         // must be OBJECT, or the object cannot be deserialized
-        assert ((field->getSType() != STI_OBJECT) ||
+        SerializedTypeID const sType {field->getSType()};
+        assert ((sType != STI_OBJECT) ||
             (field->getFName().fieldType == STI_OBJECT));
         field->addFieldID (s);
         field->add (s);
-        if (dynamic_cast<const STArray*> (field) != nullptr)
-            s.addFieldID (STI_ARRAY, 1);
-        else if (dynamic_cast<const STObject*> (field) != nullptr)
-            s.addFieldID (STI_OBJECT, 1);
+        if (sType == STI_ARRAY || sType == STI_OBJECT)
+            s.addFieldID (sType, 1);
     }
 }
 
 std::vector<STBase const*>
-STObject::getSortedFields (STObject const& objToSort)
+STObject::getSortedFields (
+    STObject const& objToSort, WhichFields whichFields)
 {
     std::vector<STBase const*> sf;
-    sf.reserve (objToSort.getCount ());
+    sf.reserve (objToSort.getCount());
 
     // Choose the fields that we need to sort.
     for (detail::STVar const& elem : objToSort.v_)
     {
-        // Pick out the fields and sort them.
         STBase const& base = elem.get();
-        if ((base.getSType () != STI_NOTPRESENT) &&
-            base.getFName ().shouldInclude (true))
+        if ((base.getSType() != STI_NOTPRESENT) &&
+            base.getFName().shouldInclude (whichFields))
         {
             sf.push_back (&base);
         }
     }
 
     // Sort the fields by fieldCode.
-    std::sort (sf.begin (), sf.end (),
-        [] (STBase const* a, STBase const* b) -> bool
+    std::sort (sf.begin(), sf.end(),
+        [] (STBase const* lhs, STBase const* rhs) -> bool
```

Comment:
Do we need the trailing return type for the compare lambda? I think we can get away with it.